### PR TITLE
Refactor: organize ethers-providers

### PR DESCRIPTION
**Pull request by prestwich**
Depends on #2122 

Cleans up the ethers-providers crate by organizing modules into a more clean layout.

TODONES:
- ~~update docs for existing modules~~
- ~~ensure all modules are documented~~

### CODE REVIEW

**Comment by gakonst** [comment]
concept ack, let's merge once the base is merged

**Comment by gakonst** [comment]
needs rebase since we merged your middleware errors PR. lgtm otherwise and supportive

**Comment by prestwich** [comment]
rebased and gitpilled

File: ethers-contract/tests/it/contract.rs [hunk 1]
```diff
@@ -24,13 +26,20 @@ mod eth_tests {
 
     #[derive(Debug)]
     pub struct MwErr<M: Middleware>(M::Error);
-    impl<M> ethers_providers::FromErr<M::Error> for MwErr<M>
+
+    impl<M: Middleware> MiddlewareError for MwErr<M>
     where
         M: Middleware,
```
Comment (DaniPopes) [inline ethers-contract/tests/it/contract.rs:32]: duplicate bound

**Comment by prestwich** [comment]
docs suck less now, this should be ready to go :)

File: ethers-providers/src/stream/watcher.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
+#![allow(clippy::return_self_not_must_use)]
```
Comment (DaniPopes) [inline ethers-providers/src/stream/watcher.rs:1]: I think this is useless?

File: examples/providers/examples/custom.rs [hunk 3]
```diff
@@ -23,6 +23,23 @@ pub enum WsOrIpcError {
...
+
+    fn as_serde_error(&self) -> Option<&serde_json::Error> {
+        match self {
+            WsOrIpcError::Ws(WsClientError::JsonError(e)) => Some(e),
+            WsOrIpcError::Ipc(IpcError::JsonError(e)) => Some(e),
+            _ => None,
+        }
+    }
+}
+
```
Comment (DaniPopes) [inline examples/providers/examples/custom.rs:42]: Needs extra explanation in doc on line 10

**Review by DaniPopes — CHANGES_REQUESTED** [review]
Very based. Nits

File: ethers-providers/src/stream/watcher.rs (hunk 2 above)
Comment (prestwich) [reply ethers-providers/src/stream/watcher.rs:1]: it was already there, so i left it without thinking about it. i'll check if removing it is reasonable

**Review by DaniPopes — APPROVED** [review]

**Review by gakonst — APPROVED** [review]

**Pull request merged**